### PR TITLE
IGVF-420-add-backend-dashboard

### PR DESCRIPTION
**Pull request by ottojolanki**

### CODE REVIEW

File: cdk/infrastructure/constructs/backend.py [hunk 1]
```diff
@@ -119,6 +132,13 @@ def __init__(
         self._run_batch_upgrade_automatically()
         self._run_update_mapping_automatically()
         self._add_alarms()
+        self._add_dashboard()
+
+    def _define_log_driver_for_pyramid_container(self) -> None:
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:137]: Can you put this right above where application container gets defined?

File: cdk/infrastructure/constructs/backend.py [hunk 2]
```diff
@@ -101,6 +113,7 @@ def __init__(
         self._generate_session_secret()
         self._define_docker_assets()
         self._define_domain_name()
+        self._define_log_driver_for_pyramid_container()
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:116]: Group next to application container.

File: cdk/infrastructure/constructs/backend.py [hunk 3]
```diff
@@ -363,3 +380,15 @@ def _add_alarms(self) -> None:
                 fargate_service=self.fargate_service
             )
         )
+
+    def _add_dashboard(self) -> None:
+        aws_logs = cast(AwsLogDriver, self.pyramid_log_driver)
+        log_group = cast(LogGroup, aws_logs.log_group)
+        self.dashboard = BackendDashboard(
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:387]: I don't think we need to assign this to `self.dashboard`?

File: cdk/infrastructure/constructs/backend.py [hunk 4]
```diff
@@ -363,3 +380,15 @@ def _add_alarms(self) -> None:
...
+            self,
+            'PyramidDashBoard',
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:389]: 'BackendDashboard'

File: cdk/infrastructure/constructs/backend.py (hunk 1 above)
Comment (keenangraham) [reply cdk/infrastructure/constructs/backend.py:137]: `_define_log_driver_for_pyramid_container`
Can we call it `_define_log_driver_for_application_container` to be consistent?

File: cdk/infrastructure/constructs/backend.py [hunk 5]
```diff
@@ -77,6 +88,7 @@ class Backend(Construct):
 
     props: BackendProps
     postgres: PostgresConstruct
+    pyramid_log_driver: LogDriver
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:91]: Can we call this `application_log_driver` to be consistent?

File: cdk/infrastructure/constructs/dashboards/backend.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            left=[response_5xx_metric],
+            left_y_axis=YAxisProps(show_units=False),
+            period=cdk.Duration.seconds(60),
+        )
+        self._widgets.append(response_5xx_widget)
+
+    def _define_dashboard(self) -> None:
+        self.dashboard = Dashboard(
+            self,
+            'BackendDashBoard'
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/dashboards/backend.py:209]: 'Dashboard'

**Review by keenangraham — CHANGES_REQUESTED** [review]
Looks good, just some organizational stuff.

File: cdk/infrastructure/constructs/backend.py [hunk 7]
```diff
@@ -19,11 +20,21 @@
 from aws_cdk.aws_secretsmanager import Secret as SMSecret
 from aws_cdk.aws_secretsmanager import SecretStringGenerator
 
+from aws_cdk.aws_cloudwatch import Dashboard
+from aws_cdk.aws_cloudwatch import GraphWidget
+from aws_cdk.aws_cloudwatch import LogQueryWidget
+from aws_cdk.aws_cloudwatch import YAxisProps
```
[the remark this answers is not in the document]
Comment (ottojolanki) [reply cdk/infrastructure/constructs/backend.py:26]: thanks. forgot to remove after putting dashboard to separate module.

**Review by keenangraham — APPROVED** [review]
LGTM.

File: cdk/infrastructure/constructs/backend.py [hunk 8]
```diff
@@ -363,3 +375,15 @@ def _add_alarms(self) -> None:
                 fargate_service=self.fargate_service
             )
         )
+
+    def _add_dashboard(self) -> None:
+        aws_logs = cast(AwsLogDriver, self.application_log_driver)
+        log_group = cast(LogGroup, aws_logs.log_group)
+        dashboard = BackendDashboard(
+            self,
+            'BackendDashBoard',
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/backend.py:384]: `BackendDashboard`

File: cdk/infrastructure/constructs/dashboards/backend.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+            left=[response_5xx_metric],
+            left_y_axis=YAxisProps(show_units=False),
+            period=cdk.Duration.seconds(60),
+        )
+        self._widgets.append(response_5xx_widget)
+
+    def _define_dashboard(self) -> None:
+        self.dashboard = Dashboard(
+            self,
+            'DashBoard'
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/dashboards/backend.py:209]: `Dashboard`

File: cdk/infrastructure/constructs/dashboards/backend.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        )
```
Comment (keenangraham) [inline cdk/infrastructure/constructs/dashboards/backend.py:210]: I think we might need to define `dashboard_name` (https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_cloudwatch/Dashboard.html) here so we can tell which backend it's reporting on. Something like `dashboard_name = f'{self.props.config.common.project_name}-{self.props.config.branch}-backend'`.

**Review by keenangraham — CHANGES_REQUESTED** [review]

**Review by keenangraham — APPROVED** [review]

**Review by keenangraham — APPROVED** [review]
LGTM.

**Pull request merged**